homeassistant/components/vicare/number.py

```diff
@@ -0,0 +1,164 @@
+    ViCareNumberEntityDescription(
+        key="heating curve slope",
+        name="Heating curve slope",
+        icon="mdi:slope-uphill",
+        entity_category=EntityCategory.CONFIG,
+        value_getter=lambda api: api.getHeatingCurveSlope(),
+        value_setter=lambda api, slope: (
+            api.setHeatingCurve(api.getHeatingCurveShift(), slope)
+        ),
+        native_min_value=0.2,
+        native_max_value=3.5,
+        native_step=0.1,
+    ),
+)
+
+
+def _build_entity(
+    name: str,
+    vicare_api: PyViCareHeatingDeviceWithComponent,
+    device_config: PyViCareDeviceConfig,
+    entity_description: ViCareNumberEntityDescription,
+) -> ViCareNumber | None:
+    """Create a ViCare number entity."""
+    _LOGGER.debug("Found device %s", name)
+    if is_supported(name, entity_description, vicare_api):
+        return ViCareNumber(
+            name,
+            vicare_api,
+            device_config,
+            entity_description,
+        )
+    return None
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Create the ViCare number devices."""
+    api = hass.data[DOMAIN][config_entry.entry_id][VICARE_API]
+
+    entities: list[ViCareNumber] = []
+    try:
+        for circuit in api.circuits:
+            suffix = ""
+            if len(api.circuits) > 1:
+                suffix = f" {circuit.id}"
+            for description in CIRCUIT_ENTITY_DESCRIPTIONS:
+                entity = await hass.async_add_executor_job(
+                    _build_entity,
+                    f"{description.name}{suffix}",
+                    circuit,
+                    hass.data[DOMAIN][config_entry.entry_id][VICARE_DEVICE_CONFIG],
+                    description,
+                )
+                if entity is not None:
+                    entities.append(entity)
+    except PyViCareNotSupportedFeatureError:
+        _LOGGER.debug("No circuits found")
+
+    async_add_entities(entities)
+
+
+class ViCareNumber(ViCareEntity, NumberEntity):
+    """Representation of a ViCare number."""
+
+    entity_description: ViCareNumberEntityDescription
+
+    def __init__(
+        self,
+        name: str,
+        api: PyViCareHeatingDeviceWithComponent,
+        device_config: PyViCareDeviceConfig,
+        description: ViCareNumberEntityDescription,
+    ) -> None:
+        """Initialize the number."""
+        super().__init__(device_config, api, description.key)
+        self.entity_description = description
+        self._attr_name = name
+
+    @property
+    def available(self) -> bool:
+        """Return True if entity is available."""
+        return self._attr_native_value is not None
+
+    def set_native_value(self, value: float) -> None:
+        """Set new value."""
+        if self.entity_description.value_setter:
+            self.entity_description.value_setter(self._api, value)
+        self.async_write_ha_state()
```

Comment:
We're not allowed to use the async API from sync context. Use `self.schedule_update_ha_state` to update state from sync context.

Reply:
Thanks for spotting that! https://github.com/home-assistant/core/pull/104687